File: cpp/prime.cpp
#include <iostream>
#include <cmath>
#include <chrono>

using namespace std;

int main()
{
    int last = 10000;
    int found = 4;   // we already know 2, 3, 5, 7
    const int arraylength = (int)(last / log(last));
    int primes[10000] = {3, 5, 7};
    cout << "Calculating prime numbers until " << last << endl;
    auto start = chrono::high_resolution_clock::now();
    for (int number = 11; number < last; number += 2)
    {
        int prime = 1;
        for (int divider = 3; divider < (int)(sqrt(number)) + 1; divider += 2)
        {
            if (number % divider == 0)
            {
                prime = 0;
                break;
            }
        }
        if (prime ==  1)
        {
            found += 1;
        }
    }
    auto end = chrono::high_resolution_clock::now();
    double time_taken = chrono::duration_cast<chrono::nanoseconds>(end - start).count();
    time_taken *= 1e-9;
    cout << "Found " << found << " prime numbers." << endl;
    cout << "This took " << fixed << time_taken << " seconds." << endl;
}

File: rust/prime.rs
// Calculating primes to {last} in Rust v3.0.2023

use std::time::{Instant};

fn is_prime(number: i32) -> i32 {
    let mut prime = 1;
    for divider in (3..((number as f32).sqrt() as i32) + 1).step_by(2)
    {
        if number % divider == 0
        {
            prime = 0;
            break;
        }
    }
    return prime;
}

// This is the main function.
fn main() {
    let last  = 100000000;
    let mut found = 4;   // we already know 2, 3, 5, 7
    // let primes = [2, 3, 5, 7];
    println!("Calculating prime numbers until {last} in Rust v3.0.2023");
    let start = Instant::now();
    for number in (11..last).step_by(2)
    {
        found += is_prime(number);
    }
    let duration = start.elapsed();
    println!("Found {found} prime numbers.");
    println!("This took {:?} seconds.", duration);    
}
